### PR TITLE
Add some validation to ListingTable to verify that at least one file exists

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/1736

# Rationale for this change

It is very frustrating to get incorrect query results when trying to query csv files with a non-csv extension such as `.tbl` or `.tsv`.

I at least get an error now but on query execution not when registering tables. Not sure if this is what we want?

```
❯ create external table partsupp (ps_partkey int, ps_suppkey int, ps_availqty int, ps_supplycost float, ps_comment varchar) stored as csv with delimiter '|' location '../benchmarks/data/partsupp.tbl';
0 rows in set. Query took 0.002 seconds.
❯ select * from partsupp;
Plan("No files found with extension .csv")
```

# What changes are included in this PR?

Add some validation that files exists

# Are there any user-facing changes?

### CODE REVIEW

**Comment by andygrove** [comment]
@tustvold Would be good to get some feedback on the approach here before I add tests.

File: datafusion/core/src/datasource/listing/table.rs [hunk 1]
```diff
@@ -443,6 +443,20 @@ impl ListingTable {
         let store = ctx
             .runtime_env
             .object_store(&self.table_paths.get(0).unwrap())?;
+
+        let mut count = 0;
+        for table_path in &self.table_paths {
+            let stream = table_path.list_all_files(store.as_ref(), &self.options.file_extension);
+            let mut all_files = Box::pin(stream.fuse());
+            while let Some(res) = all_files.next().await {
```
Comment (tustvold) [inline datafusion/core/src/datasource/listing/table.rs:451]: Why not short-circuit on the first file?

**Review by tustvold — COMMENTED** [review]
I don't object to this change, however, I do wonder if we should just remove the filtering based on file extension. This would mean:

* If the path doesn't exist, ListingTableUrl will return an error
* If the file is the wrong format, you will get such an error

It seems strange to me that you specify the file is a CSV file, and then it also cares about the extension.

If the concern is multiple formats in the same directory, we have glob expressions as an escape hatch for that (and I would also question the data partitioning as it will confuse many tools)

**Comment by andygrove** [comment]
> I do wonder if we should just remove the filtering based on file extension

I would much prefer that. I will update this PR to use that approach instead.

**Comment by andygrove** [comment]
> It seems strange to me that you specify the file is a CSV file, and then it also cares about the extension.

One of the motivations may have been to handle the Spark case, where directories may contain a `_SUCCESS` file alongside the actual data files.

**Pull request closed without merging**